### PR TITLE
[Driver][SYCL][FPGA] Add output report folder when building from AOCR…

### PR DESCRIPTION
**Pull request by mdtoguchi**
… archive

When building for FPGA and the input is an AOCR archive, the output
report folder was not being added to the offline compilation step.

### CODE REVIEW

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 1]
```diff
@@ -311,6 +311,7 @@ void SYCL::fpga::BackendCompiler::ConstructJob(
 
   InputInfoList ForeachInputs;
   InputInfoList FPGADepFiles;
+  StringRef createdReportName;
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp:314]: (Optional) suggestion for renaming as per LLVM standards along the way:
```suggestion
  StringRef CreatedReportName;
```

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 2]
```diff
@@ -324,6 +325,21 @@ void SYCL::fpga::BackendCompiler::ConstructJob(
...
+    // Check for any AOCR input, if found use that as the project report name
+    StringRef Ext(llvm::sys::path::extension(Filename));
+    if (Ext.empty())
+      continue;
+    if (getToolChain().LookupTypeForExtension(Ext.drop_front()) ==
+        types::TY_FPGA_AOCR) {
+      // Keep the base of the .aocr file name.  Input file is a temporary,
+      // so we are stripping off the additional naming information for a
+      // cleaner name.
+      SmallString<128> NameBase(Filename.substr(0, Filename.length() - 12));
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp:337]: I personally find this magic integer constant really difficult to decode. Could it be set into a named constant variable? Or, perhaps, calculated based on the filename + FS specifics?

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 3]
```diff
@@ -324,6 +325,21 @@ void SYCL::fpga::BackendCompiler::ConstructJob(
...
+      NameBase.append(".aocr");
+      llvm::sys::path::replace_extension(NameBase, "prj");
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp:339]: Is there a reason not to prefer the following?
```suggestion
      NameBase.append(".prj");
```

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 1 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: Done

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 2 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp:338]: Added some additional doc on how the number is formulated.

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 3 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: hah - good catch.  I have streamlined based on your suggestion.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 4]
```diff
@@ -324,6 +325,21 @@ void SYCL::fpga::BackendCompiler::ConstructJob(
...
+    StringRef Ext(llvm::sys::path::extension(Filename));
+    if (Ext.empty())
+      continue;
+    if (getToolChain().LookupTypeForExtension(Ext.drop_front()) ==
+        types::TY_FPGA_AOCR) {
+      // Keep the base of the .aocr file name.  Input file is a temporary,
+      // so we are stripping off the additional naming information for a
+      // cleaner name.  The length is the being stripped from the name is the
+      // added temporary string and the extension:  input_file-XXXXXX.aocr
+      SmallString<128> NameBase(Filename.substr(0, Filename.length() - 12));
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp:338]: I would suggest something like this with the aim of guarding us humans from manual counting:
```suggestion
      // cleaner name. The length is the being stripped from the name is the
      // added temporary string and the extension:
      constexpr std::string_view SuffixFormat = "input_file-XXXXXX.aocr";
      SmallString<128> NameBase(Filename.substr(0, Filename.length() - SuffixFormat.length()));
```

Comment (AGindinson) [reply clang/lib/Driver/ToolChains/SYCL.cpp:338]: So yeah, perhaps just this. I don't know if our `StringRef` grants the same compile-time fanciness, but should still be fine for the minor purpose.
```suggestion
      // cleaner name. The length is the being stripped from the name is the
      // added temporary string and the extension:
      StringRef SuffixFormat("input_file-XXXXXX.aocr");
      SmallString<128> NameBase(Filename.substr(0, Filename.length() - SuffixFormat.size()));
```
EDIT: string_view is C++17, though... I'll update the comment if I find a synonym in C++11.

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: I like it - done.

**Review by AGindinson — DISMISSED** [review]
LGTM, thanks!

**Review by AGindinson — APPROVED** [review]

**Pull request merged**